bionlp/preprocess/parser.py

```diff
@@ -0,0 +1,96 @@
+    # for Brat, we only want to regard the input files with file extension .ann
+    def parse_datasets(self, input_dir, output_dir):
+        # this methodology only expects files in the top level of the given input directory, not subdirs
+        _, _, input_files = next(os.walk(input_dir))
+        for input_file in input_files:
+            if input_file.endswith(".ann"):
+                input_file_abs_path = os.path.join(input_dir, input_file)
+                output_file_abs_path = os.path.join(output_dir, os.path.splitext(input_file)[0] + ".json")
+                self.parse_file(input_file_abs_path, output_file_abs_path)
+
+    def parse_file(self, input_file, output_file):
+        total_tags_list = []
+        with open(input_file, "r", encoding="utf8") as f_in:
+            tag_chain_beginning_index = None
+            tag_chain_end_index = None
+            tag_chain_token_values = ""
+            tag_chain_tag = None
+            tag_chain_id = -1
+            for line in f_in.readlines():
+                cleaned_line = line.rstrip()
+                brat_index, tag_and_indices, token_value = cleaned_line.split("\t")
+                tag_string, indices = self.split_tag_and_indices(tag_and_indices)
+                tag = IOB2_Tag(tag_string)
+                earliest_start_index, latest_end_index = self.get_earliest_start_and_latest_end_index(indices)
+                if IOB2_Tag.is_tag_continuation_of_beginning_tag(tag_chain_tag, tag):
+                    tag_chain_end_index = latest_end_index
+                    tag_chain_token_values = " ".join([tag_chain_token_values, token_value])
+                else:
+                    # if a previous tag chain exists, append it to the total tags list
+                    if tag_chain_tag is not None:
+                        self.append_tag_chain_to_total_list(total_tags_list, tag_chain_beginning_index,
+                            tag_chain_end_index, tag_chain_token_values, tag_chain_tag, tag_chain_id)
+
+                    # begin new tag chain
+                    tag_chain_beginning_index = earliest_start_index
+                    tag_chain_end_index = latest_end_index
+                    tag_chain_token_values = token_value
+                    tag_chain_tag = tag
+                    tag_chain_id += 1
+
+            # after the final line has been read, write the final tag chain into the total tags list
+            self.append_tag_chain_to_total_list(total_tags_list, tag_chain_beginning_index,
+                            tag_chain_end_index, tag_chain_token_values, tag_chain_tag, tag_chain_id)
+
+        # write the total tags list into the output file
+        with open(output_file, "w", encoding="utf8") as f_out:
+            json.dump(total_tags_list, f_out, ensure_ascii=False)
+
+    def append_tag_chain_to_total_list(self, total_tags_list, tag_chain_beginning_index, tag_chain_end_index, tag_chain_token_values, tag_chain_tag, tag_chain_id):
```

Comment:
I'm not convinced that declaring this function makes the code cleaner. I think its easier to append the tag chain in lines 49 and 60 directly